openflow13/match.go

```diff
@@ -1323,3 +1330,100 @@ func NewTunnelIpv4DstField(tunnelIpDst net.IP, tunnelIpDstMask *net.IP) *MatchFi
 
 	return f
 }
+
+// ARP Host Address field message, used by arp_sha and arp_tha match
+type ArpXHaField struct {
+	arpHa net.HardwareAddr
+}
+
+func (m *ArpXHaField) Len() uint16 {
+	return 6
+}
+func (m *ArpXHaField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, m.Len())
+	copy(data, m.arpHa)
+	return
+}
+
+func (m *ArpXHaField) UnmarshalBinary(data []byte) error {
+	if len(data) < int(m.Len()) {
+		return errors.New("The byte array has wrong size to unmarshal ArpXHaField message")
+	}
+	copy(m.arpHa, data[:6])
+	return nil
+}
+
+func NewArpThaField(arpTha net.HardwareAddr) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_ARP_THA
+	f.HasMask = false
+
+	arpThaField := new(ArpXHaField)
+	arpThaField.arpHa = arpTha
+	f.Value = arpThaField
+	f.Length = uint8(arpThaField.Len())
+	return f
+}
+
+func NewArpShaField(arpSha net.HardwareAddr) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_ARP_SHA
+	f.HasMask = false
+
+	arpXHAField := new(ArpXHaField)
+	arpXHAField.arpHa = arpSha
+	f.Value = arpXHAField
+	f.Length = uint8(arpXHAField.Len())
+	return f
+}
+
+// ARP Protocol Address field message, used by arp_spa and arp_tpa match
+type ArpXPaField struct {
+	ArpPa net.IP
+}
+
+func (m *ArpXPaField) Len() uint16 {
+	return 4
+}
+
+func (m *ArpXPaField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, m.Len())
+	copy(data, m.ArpPa.To4())
+	return
+}
+
+func (m *ArpXPaField) UnmarshalBinary(data []byte) error {
+	if len(data) < int(m.Len()) {
+		return errors.New("The byte array has wrong size to unmarshal ArpXPaField message")
+	}
+	m.ArpPa = net.IPv4(data[0], data[1], data[2], data[3])
+	return nil
+}
+
+func NewArpTpaField(arpTpa net.IP) *MatchField {
```

Comment:
I think this is the same as the next function. I suggest refactoring in terms of a new helper.

Reply:
As existent code in libOpenflow is using this style, every field is constructed in a specific function, and all code is similar. So I just follow the style. I think it could be optimized in another MR if necessary.